### PR TITLE
Use client settings in repository-gcs

### PR DESCRIPTION
**Pull request by tlrx**
Similarly to what has been done for s3 and azure, this pull request removes
the repository settings `application_name` and `connect/read_timeout`
in favour of settings at the client level. 

It introduces a `GoogleCloudStorageClientSettings`
class (similar to `S3ClientSettings`) and a bunch of unit tests for that,
it aligns the documentation to be more coherent with the S3 one, it
documents the connect/read timeouts that were not documented at all 
and also adds a new client setting that allows to define a custom `endpoint`.

### CODE REVIEW

**Comment by tlrx** [comment]
@rjernst Would you like to review this please? I'd like to move forward on this.

File: docs/plugins/repository-gcs.asciidoc [hunk 1]
```diff
@@ -116,36 +116,90 @@ PUT _snapshot/my_gcs_repository
...
+    "bucket": "my_bucket",
+    "client": "my_alternate_client"
+  }
+}
+----
+// CONSOLE
+// TEST[skip:we don't have gcs setup while testing this]
+
+Some settings are sensitive and must be stored in the
+{ref}/secure-settings.html[elasticsearch keystore]. This is the case of the service account file:
```
Comment (ywelsch) [inline docs/plugins/repository-gcs.asciidoc]: the case *for* the ...?

File: docs/plugins/repository-gcs.asciidoc [hunk 2]
```diff
@@ -116,36 +116,90 @@ PUT _snapshot/my_gcs_repository
...
+are marked as `Secure`.
 
+`credentials_file`::
+
+    The service account file that is used to authenticate to the Google Cloud Storage service. (Secure)
+
+`endpoint`::
+
+    The Google Cloud Storage service endpoint to connect to. This will be automatically
+    figured out by the Google Cloud Storage client but can be specified explicitly.
```
Comment (ywelsch) [inline docs/plugins/repository-gcs.asciidoc]: automatically determined by the ...

File: docs/reference/migration/migrate_7_0/plugins.asciidoc [hunk 3]
```diff
@@ -12,3 +12,7 @@ You must set it per azure client instead. Like `azure.client.default.timeout: 10
 
 See {plugins}/repository-azure-usage.html#repository-azure-repository-settings[Azure Repository settings].
 
+==== Google Cloud Storage Repository plugin
+
+* The repository settings `application_name`, `connect_timeout` and `read_timeout` have been removed and
+must now be specified in the client settings instead.
```
Comment (ywelsch) [inline docs/reference/migration/migrate_7_0/plugins.asciidoc]: link to the docs here?

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStoragePlugin.java [hunk 4]
```diff
@@ -112,15 +108,19 @@
         });
     }
 
-    private final Map<String, GoogleCredential> credentials;
+    private final Map<String, GoogleCloudStorageClientSettings> clientsSettings;
```
Comment (ywelsch) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStoragePlugin.java]: clientsSettings -> clientSettings?

File: docs/plugins/repository-s3.asciidoc [hunk 5]
```diff
@@ -36,7 +36,7 @@ PUT _snapshot/my_s3_repository
 
 The client used to connect to S3 has a number of settings available. Client setting names are of
 the form `s3.client.CLIENT_NAME.SETTING_NAME` and specified inside `elasticsearch.yml`.  The
-default client name looked up by an s3 repository is called `default`, but can be customized
+default client name looked up by a `s3` repository is called `default`, but can be customized
```
Comment (ywelsch) [inline docs/plugins/repository-s3.asciidoc]: Maybe instead "The default client name used by the "s3" repository is called `default` ...

**Review by ywelsch — APPROVED** [review]
LGTM (some super minor nits)

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStoragePlugin.java (hunk 4 above)
Comment (tlrx) [reply plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStoragePlugin.java]: There can be multiple client settings keyed by their name so I chose the plural form :)

File: docs/plugins/repository-gcs.asciidoc [hunk 6]
```diff
@@ -116,36 +116,90 @@ PUT _snapshot/my_gcs_repository
...
+----
+// CONSOLE
+// TEST[skip:we don't have gcs setup while testing this]
+
+Some settings are sensitive and must be stored in the
+{ref}/secure-settings.html[elasticsearch keystore]. This is the case for the service account file:
+
+[source,sh]
+----
+bin/elasticsearch-keystore add gcs.client.default.credentials_file
```
Comment (rjernst) [inline docs/plugins/repository-gcs.asciidoc]: The subcommand is `add-file`, not `add`.

File: docs/plugins/repository-gcs.asciidoc [hunk 7]
```diff
@@ -116,36 +116,90 @@ PUT _snapshot/my_gcs_repository
...
+bin/elasticsearch-keystore add gcs.client.default.credentials_file
```
Comment (rjernst) [inline docs/plugins/repository-gcs.asciidoc]: And this line seems to be duplicated?

**Review by rjernst — APPROVED** [review]
LGTM

File: docs/plugins/repository-gcs.asciidoc (hunk 6 above)
Comment (tlrx) [reply docs/plugins/repository-gcs.asciidoc]: Good catch

**Comment by tlrx** [comment]
Thanks @ywelsch and @rjernst

**Comment by tlrx** [comment]
This change has been backported to 6.x in 240aa34cb4fab36229f308d3fe7dcdb641a30662 along with a deprecation log and updated migration documentation.

**Pull request merged**